jfix-zookeeper/src/main/java/ru/fix/zookeeper/lock/PersistentExpiringLockManager.java

```diff
@@ -148,36 +156,171 @@ public void release(LockIdentity lockId) {
-                    lockContainer.lock.close();
-                } catch (Exception e) {
-                    logger.error("Failed to close lock with lockId={}", lockId, e);
+                    globalLock.lock();
+                    final ProcessingLockResult processingLockResult;
+                    if (locks.containsKey(lockId)) {
+                        processingLockResult = lockProcessor.apply(lockId, lockContainer.lock);
+                    } else {
+                        processingLockResult = ALREADY_REMOVED;
+                    }
+                    applyProcessingLockResult(lockId, processingLockResult);
+                    lockResultPostProcessor.apply(
+                            lockId, lockContainer.prolongationFailedListener, processingLockResult
+                    );
+                } finally {
+                    globalLock.unlock();
                 }
+            });
+        }
+
+        /**
+         * clear container without closing content in {@link #locks}
+         * should be closed on the upper level!
+         */
+        @Override
+        public void close() {
+            locks.clear();
+        }
+
+        private void applyProcessingLockResult(
+                LockIdentity lockId,
+                ProcessingLockResult processingLockResult
+        ) {
+            switch (processingLockResult) {
+                case REMOVE_LOCK_FROM_CONTAINER:
+                    locks.remove(lockId);
+                    break;
+                case KEEP_LOCK_IN_CONTAINER:
+                    // do nothing
+                    break;
+                case ALREADY_REMOVED:
+                    logger.trace("Lock with lockId {} already has been removed by other thread", lockId);
+                    break;
+                default:
+                    throw new IllegalStateException(
+                            "Unexpected value for processingLockResult " + processingLockResult
+                    );
             }
-        });
-        locks.clear();
-        lockProlongationScheduler.close();
+        }
+
+        private static class LockContainer {
+            public final PersistentExpiringDistributedLock lock;
+            public final LockProlongationFailedListener prolongationFailedListener;
+
+            public LockContainer(
+                    PersistentExpiringDistributedLock lock,
+                    LockProlongationFailedListener prolongationFailedListener
+            ) {
+                this.lock = lock;
+                this.prolongationFailedListener = prolongationFailedListener;
+            }
+        }
+
+        @FunctionalInterface
+        public interface LockResultPostProcessor {
+            void apply(
+                    LockIdentity lockId,
+                    LockProlongationFailedListener lockListener,
+                    ProcessingLockResult processingLockResult
+            );
+        }
+
+        enum ProcessingLockResult {
```

Comment:
This is codestyle, but! I really don't think this is a good idea to create such enum. What does it mean? It contains "possible internal answers of local functions, all in one pile". Therefore, instead of dividing classes by their responsibilities we've conjoined them with common answers class. I think it will be a good idea to move ActiveLocksContainer out to a separate file. And try to treat it (in mind) as if it could be a part of some public api. It will help you (and others) to see what does it do and what - doesn't.

Reply:
moved to upper level and rewrite container in kotlin

Reply:
Like for kotlin!
Now it is much easier to see all ActiveLocksContainer's API. It consists of:
putLock
contains
getLockState
processAllLocks
close

Let's check out `processAllLocks`, it looks dangerous because get's two callbacks (with 2 and 3 params, whoa). It also returns custom enum as a result state. 
On the first glance it is used from two places - close (here it used twice) and regular prolongation. So maybe it shouldn't be a part of ActiveLocksContainer's public API? Maybe we have two much more concrete functions here - `prolong single lock` (or maybe `attempt to prolong all locks`) and `close all locks`, which looks like duplicate for existing `close()` method?

Reply:
after discussion come to next resolution:  
- stay with strict container responsibility without knowing about prolonging locks
- syncing only remove/iterate operations
- in the case of low performance -- synchronize by LockContainer wrapper
- in the beautiful world of future rewrite container to implementaiton of ConcurrentMap interface